todo/operations/csv_importer.py

```diff
@@ -0,0 +1,185 @@
+
+    def validate_row(self, row):
+        """Perform data integrity checks and set default values. Returns a valid object for insertion, or False.
+        Errors are stored for later display."""
+
+        row_errors = []
+
+        # #######################
+        # Task creator must exist
+        if not row.get("Created By"):
+            msg = f"Missing required task creator."
+            row_errors.append(msg)
+
+        created_by = get_user_model().objects.filter(username=row.get("Created By"))
+        if created_by.exists():
+            creator = created_by.first()
+        else:
+            creator = None
+            msg = f"Invalid task creator {row.get('Created By')}"
+            row_errors.append(msg)
+
+        # #######################
+        # If specified, Assignee must exist
+        if row.get("Assigned To"):
+            assigned = get_user_model().objects.filter(username=row.get("Assigned To"))
+            if assigned.exists():
+                assignee = assigned.first()
+            else:
+                msg = f"Missing or invalid task assignee {row.get('Assigned To')}"
+                row_errors.append(msg)
+        else:
+            assignee = None  # Perfectly valid
+
+        # #######################
+        # Group must exist
+        try:
+            target_group = Group.objects.get(name=row.get("Group"))
+        except Group.DoesNotExist:
+            msg = f"Could not find group {row.get('Group')}."
+            row_errors.append(msg)
+
+        # #######################
+        # Task creator must be in the target group
+        if creator and target_group not in creator.groups.all():
+            msg = f"{creator} is not in group {target_group}"
+            row_errors.append(msg)
+
+        # #######################
+        # Assignee must be in the target group
+        if assignee and target_group not in assignee.groups.all():
+            msg = f"{assignee} is not in group {target_group}"
+            row_errors.append(msg)
+
+        # #######################
+        # Task list must exist in the target group
+        try:
+            tasklist = TaskList.objects.get(name=row.get("Task List"), group=target_group)
+            row["Task List"] = tasklist
+        except TaskList.DoesNotExist:
+            msg = f"Task list {row.get('Task List')} in group {target_group} does not exist"
+            row_errors.append(msg)
+
+        # #######################
+        # Validate Due Date
+        dd = row.get("Due Date")
+        if dd:
+            try:
+                row["Due Date"] = datetime.datetime.strptime(dd, "%Y-%m-%d")
+            except ValueError:
+                msg = f"Could not convert Due Date {dd} to python date"
+                row_errors.append(msg)
+        else:
+            row["Created Date"] = None  # Override default empty string '' value
+
+        # #######################
+        # Validate Created Date
+        cd = row.get("Created Date")
```

Comment:
is duplicating this code snippet ok ? it seems to be the same thing as the lines just above

Reply:
Yes, I'll consolidate these when I convert the snippets to independent functions.

Reply:
Now consolidated. Decided not to break out a bunch of different validate methods due to interdendencies, but did break out the date validator.